TFNetworkRecLayer.py

```diff
@@ -1244,31 +1242,35 @@ def get_output(self, rec_layer):
       # if not self.input_data and self.network.search_flag:
       #   assert not have_known_seq_len  # at least for the moment
 
-      # TODO: Better check for train_flag.
-      # Maybe more generic via sampling options later.
-      y_ta = None
-      if rec_layer.target and ((
-            rec_layer.network.train_flag is not False and not self.parent_net.search_flag) or rec_layer._cheating):
-        # TODO check subnet, which extern data keys are used...
-        y_data = rec_layer.network.get_extern_data(rec_layer.target, mark_data_key_as_used=True)
-        y = y_data.get_placeholder_as_time_major()
-        with tf.name_scope("y_max_len"):
-          y_max_len = tf.shape(y)[0]
+      used_keys = self.net.used_data_keys.copy()
+      if rec_layer.target:
+        used_keys.add(rec_layer.target)  # we always need the target of the recurrent layer
+      for key in used_keys:
+        # TODO: Better check for train_flag.
+        # Maybe more generic via sampling options later.
+        if key == rec_layer.target and (
+            rec_layer.network.train_flag is False or self.parent_net.search_flag) and not rec_layer._cheating:
+          continue
+        data = rec_layer.network.get_extern_data(key, mark_data_key_as_used=True)
+        data_placeholder = data.get_placeholder_as_time_major()
+        with tf.name_scope("data_max_len"):
+          data_max_len = tf.shape(data_placeholder)[0]
           if input_seq_len is not None:
-            with tf.control_dependencies([tf.assert_equal(tf.reduce_max(input_seq_len), y_max_len,
+            with tf.control_dependencies([tf.assert_equal(tf.reduce_max(input_seq_len), data_max_len,
                 ["RecLayer %r with sources %r." % (rec_layer.name, rec_layer.sources),
                  " The length of the sources (", tf.reduce_max(input_seq_len),
-                 ") differ from the length of the target (", y_max_len, ")."])]):
-              y_max_len = tf.identity(y_max_len)
-        y_ta = tf.TensorArray(
-          name="y_ta",
-          dtype=y_data.dtype,
-          element_shape=tf.TensorShape(y_data.copy_template_excluding_time_dim().batch_shape),
-          size=y_max_len,
+                 ") differ from the length of the target (", data_max_len, ")."])]):
+              data_max_len = tf.identity(data_max_len)
+        data_ta = tf.TensorArray(
+          name=key + "_ta",
+          dtype=data.dtype,
+          element_shape=tf.TensorShape(data.copy_template_excluding_time_dim().batch_shape),
+          size=data_max_len,
           infer_shape=True)
-        y_ta = y_ta.unstack(y, name="y_ta_unstack")
+        data_ta = data_ta.unstack(data_placeholder, name="{}_ta_unstack".format(key))
+        data_tensor_arrays[key] = data_ta
         if max_seq_len is None:
-          max_seq_len = y_max_len
+          max_seq_len = data_max_len
```

Comment:
This behavior will depend on the order in `used_data_keys`, which can be non-deterministic. This code should not be non-deterministic.